Make tables_without_deprecation really avoid causing deprecations instead of just ineffectively suppressing them

boazy opened the pull request:
This Pull Request also does its best to actually emulated the desired behavior of ConnectionAdapter#tables (returning only the tables, without the views) in all the supported databases (MySQL, Postgress, SQLite3) and even in unsupported databases who have the AR5 compatible methods `views` and `data_sources`.

boazy committed:
Better implementation of AbstractAdapter#tables_without_deprecation.

The existing implementation was still causing deprecation warnigs.
They were silenced, but even silenced warnings trigger spec failures
when you expect another deprecation.

boazy committed:
Properly test deprecations

boazy committed:
Test that views are not included in tables result.

ronen commented on spec/abstract_adapter_spec.rb:
```diff
@@ -20,10 +20,9 @@
-        connection.tables_without_deprecation
-      }
+    it "does not cause deprecation" do
+      expect(ActiveSupport::Deprecation).not_to receive(:warn)
```
I could have sworn I tried this and it didn't work for some reason, so I went with the cumbersome helper.  Can't imagine what the trouble was, and of course this way is much cleaner.

...though, have you verified that if you just called `connection.tables` the spec _would_ fail, at least in some cases?  (That's the problem with `expect ... not_to` tests, it's hard to know if they're actually testing anything!)

ronen commented on lib/schema_plus/compatibility/active_record/connection_adapters/abstract_adapter.rb:
```diff
@@ -3,9 +3,19 @@ module ActiveRecord
+          all_sources = (try :data_sources) || tables
+          views_only = (try :views) || []
+          all_sources - views_only
         end
```
I think it's safe to say we don't support non-supported DBs :)  

So I'd remove this method entirely, rather than keeping around code that's not tested (and likely never used).

ronen commented:
Thanks as always for the extra work on all the DB's!  

I'm OK with having `tables_without_deprecation` really return only `tables` for all versions.  The README should be updated accordingly too.

boazy committed:
Replaced connection.tables_without_deprecation with connection.tables_only

ronen commented:
Impressive effort to make and test a custom adapter!  But I still don't see why bother.  Why have that abstract code and why test it?   SchemaPlus in general won't work with non-standard, unsupported DB adapter.  Do you have a use case in mind?

It does occur to me that the file could be renamed from `abstract_adapter_spec` to `tables_only_spec.rb` and the top-level `describe` would be `describe "tables_only" ...`

boazy commented:
Yeah, it's a long stretch, but I guess I just wanted to have that covered. We could just remove the code and the test entirely.

There is a version bump to 0.2.0 by the way.

boazy committed:
Rename spec file

boazy commented:
Are we ready to integrate this? I want to move forward with schema_plus_foreign_keys

boazy committed:
Fix typo in README.md

boazy committed:
Added user_tables_only

boazy committed:
Fix test

ronen committed:
Remove support for non-supported DBs

ronen committed:
gratuitous code cleanup

ronen committed:
spec cleanup — remove unneeded test

Since latest_version is a method we just added, no specific need to test for deprecations

ronen committed:
Don’t use SchemaMonkey module in spec

It’s neat & clever to define a SchemaMonkey client to add support methods.  But I think we shouldn’t:  We’re testing additions & modifications to AR, we shouldn’t introduce additional ones as part of the testing.

Also, while the helper was required and registered in this file, the effect is actually global, which could be confusing.

ronen committed:
update README for `tables_only`

ronen committed:
use ActiveRecord::InternalMetadata.table_name rather than hardwiring `ar_internal_metadata`

ronen commented:
@boazy I took the liberty of adding a few commits with some cleanups.  Most significantly being removing the non-supported DB method.  Since none of these affect the external behavior, I'm going to take the liberty of merging these and releasing the gem, so that the other work you're doing won't continue to be blocked.

But if you disagree with or otherwise want to improve my changes, of course let me know and we can fix them in master and/or a patch release.
